File: Abstract Factory & Singleton/TestRichie/TestRichie/TestRichie.cpp
// TestRichie.cpp : Defines the entry point for the console application.
//

#include "TestRichie.h"
#include <iostream>
#include <stddef.h>

using namespace std;


#define TestRichie
#define TestAdvancedRichie
#define TestSuperRichie
#define TestUltraRichie

Richie* Richie::instance = NULL;
AdvancedRichie* AdvancedRichie::instance = NULL;
SuperRichie* SuperRichie::instance = NULL;
UltraRichie* UltraRichie::instance = NULL;


int main(int argc, char**argv)
{
	cout << "======================================" << endl;
	cout << "------------Test Richie---------------" << endl;
	cout << "-------------------------by Yu Chen---" << endl;

	cout << "======================================" << endl;

#ifdef TestRichie
	Richie* richie = (Richie*)AbstractRichie::GetConcreteRichie(RICHIE);
	Richie* r = (Richie*)AbstractRichie::GetConcreteRichie(RICHIE);
	if (richie == r) cout << "Richie has already been instantiated!" << endl;
	// test
	richie->TestCPU();
	richie->TestMMU();
	richie->TestMotherboard();
	delete richie;
#endif

#ifdef TestAdvancedRichie
	AdvancedRichie* advanced_richie = (AdvancedRichie*)AbstractRichie::GetConcreteRichie(ADVANCEDRICHIE);
	AdvancedRichie* r1 = (AdvancedRichie*)AbstractRichie::GetConcreteRichie(ADVANCEDRICHIE);
	if (advanced_richie == r1) cout << "Advanced Richie has already been instantiated!" << endl;
	// test
	advanced_richie->TestCPU();
	advanced_richie->TestMMU();
	advanced_richie->TestMotherboard();
	delete advanced_richie;
#endif

#ifdef TestSuperRichie
	SuperRichie* super_richie = (SuperRichie*)AbstractRichie::GetConcreteRichie(SUPERRICHIE);
	SuperRichie* r2 = (SuperRichie*)SuperRichie::GetConcreteRichie(ADVANCEDRICHIE);
	if (super_richie == r2) cout << "Super Richie has already been instantiated!" << endl;
	// test
	super_richie->TestCPU();
	super_richie->TestMMU();
	super_richie->TestMotherboard();
	delete super_richie;
#endif

#ifdef TestUltraRichie
	UltraRichie* ultra_richie = (UltraRichie*)AbstractRichie::GetConcreteRichie(ULTRARICHIE);
	UltraRichie* r3 = (UltraRichie*)UltraRichie::GetConcreteRichie(ADVANCEDRICHIE);
	if (ultra_richie == r3) cout << "Ultra Richie has already been instantiated!" << endl;
	// test
	ultra_richie->TestCPU();
	ultra_richie->TestMMU();
	ultra_richie->TestMotherboard();
	delete ultra_richie;
#endif
	system("pause");
	return 0;
}

AbstractRichie* AbstractRichie::GetConcreteRichie(int richie_type)
{
	switch (richie_type) {
	case RICHIE:
		return Richie::getInstance();
		break;
	case ADVANCEDRICHIE:
		return AdvancedRichie::getInstance();
		break;
	case SUPERRICHIE:
		return SuperRichie::getInstance();
		break;
	case ULTRARICHIE:
		return UltraRichie::getInstance();
		break;
	default:
		break;
	}

}


void Richie::TestCPU(void)
{
	CPUOfRichie cup_richie;
	cup_richie.test();
}

void Richie::TestMMU(void)
{
	MMUOfRichie mmu_richie;
	mmu_richie.test();
}
void Richie::TestMotherboard(void)
{
	MotherboardOfRichie motherboard_richie;
	motherboard_richie.test();
}

Richie* Richie::getInstance()
{
	if (instance == NULL)
		instance = new Richie;
	return instance;
}

void AdvancedRichie::TestCPU(void)
{
	CPUOfAdvancedRichie cup_advancedrichie;
	cup_advancedrichie.test();
}

void AdvancedRichie::TestMMU(void)
{
	MMUOfAdvancedRichie mmu_advancedrichie;
	mmu_advancedrichie.test();
}
void AdvancedRichie::TestMotherboard(void)
{
	MotherboardOfAdvancedRichie motherboard_advancedrichie;
	motherboard_advancedrichie.test();
}


AdvancedRichie* AdvancedRichie::getInstance()
{
	if (instance == NULL)
		instance = new AdvancedRichie;
	return instance;
}


void SuperRichie::TestCPU(void)
{
	CPUOfSuperRichie cup_superrichie;
	cup_superrichie.test();
}

void SuperRichie::TestMMU(void)
{
	MMUOfSuperRichie mmu_superrichie;
	mmu_superrichie.test();
}
void SuperRichie::TestMotherboard(void)
{
	MotherboardOfSuperRichie motherboard_superrichie;
	motherboard_superrichie.test();
}

SuperRichie* SuperRichie::getInstance()
{
	if (instance == NULL)
		instance = new SuperRichie;
	return instance;
}

void UltraRichie::TestCPU(void)
{
	CPUOfUltraRichie cup_ultrarichie;
	cup_ultrarichie.test();
}

void UltraRichie::TestMMU(void)
{
	MMUOfUltraRichie mmu_ultrarichie;
	mmu_ultrarichie.test();
}
void UltraRichie::TestMotherboard(void)
{
	MotherboardOfUltraRichie motherboard_ultrarichie;
	motherboard_ultrarichie.test();
}


UltraRichie* UltraRichie::getInstance()
{
	if (instance == NULL)
		instance = new UltraRichie;
	return instance;
}

void CPUOfRichie::test(void)
{
	cout << "Richie CPU" << endl;
}

void MMUOfRichie::test(void)
{
	cout << "Richie MMU" << endl;
}

void MotherboardOfRichie::test(void)
{
	cout << "Richie Motherboard" << endl;
}

void CPUOfAdvancedRichie::test(void)
{
	cout << "Advanced Richie CPU" << endl;
}

void MMUOfAdvancedRichie::test(void)
{
	cout << "Advanced Richie MMU" << endl;
}

void MotherboardOfAdvancedRichie::test(void)
{
	cout << "Advanced Richie Motherboard" << endl;
}

void CPUOfSuperRichie::test(void)
{
	cout << "Super Richie CPU" << endl;
}

void MMUOfSuperRichie::test(void)
{
	cout << "Super Richie MMU" << endl;
}

void MotherboardOfSuperRichie::test(void)
{
	cout << "Super Richie Motherboard" << endl;
}

void CPUOfUltraRichie::test(void)
{
	cout << "Ultra Richie CPU" << endl;
}

void MMUOfUltraRichie::test(void)
{
	cout << "Ultra Richie MMU" << endl;
}

void MotherboardOfUltraRichie::test(void)
{
	cout << "Ultra Richie Motherboard" << endl;
}

File: Abstract Factory & Singleton/TestRichie/TestRichie (Python)/TestRichie.py
# singleton
def singleton(cls, *args, **kw):  
    instances = {}  
    def _singleton():  
        if cls not in instances:  
            instances[cls] = cls(*args, **kw)  
        return instances[cls]  
    return _singleton  

# abstract factory     
class AbstractRichie(object):
    def TestCPU(self): pass
    def TestMMU(self): pass
    def TestMotherboard(self): pass
    def GetConcreteRichie(type):
        if type == "RICHIE": return Richie()
        if type == "ADVANCEDRICHIE": return AdvancedRichie()
        if type == "SUPERRICHIE": return SuperRichie()
        if type == "ULTRARICHIE": return UltraRichie()
        assert 0, "Bad creation: " + type
    GetConcreteRichie = staticmethod(GetConcreteRichie)

# concrete factory
@singleton 
class Richie(AbstractRichie):
    def __init__(self):
        super(AbstractRichie, self).__init__()
        self.CPU = CPUOfRichie()
        self.MMU = MMUOfRichie()
        self.Motherboard = MotherboardOfRichie()
    def TestCPU(self): self.CPU.test()
    def TestMMU(self): self.MMU.test()
    def TestMotherboard(self): self.Motherboard.test()
    
@singleton 
class AdvancedRichie(AbstractRichie):
    def __init__(self):
        super(AbstractRichie, self).__init__()
        self.CPU = CPUOfAdvancedRichie()
        self.MMU = MMUOfAdvancedRichie()
        self.Motherboard = MotherboardOfAdvancedRichie()
    def TestCPU(self): self.CPU.test()
    def TestMMU(self): self.MMU.test()
    def TestMotherboard(self): self.Motherboard.test()

@singleton 
class SuperRichie(AbstractRichie):
    def __init__(self):
        super(AbstractRichie, self).__init__()
        self.CPU = CPUOfSuperRichie()
        self.MMU = MMUOfSuperRichie()
        self.Motherboard = MotherboardOfSuperRichie()
    def TestCPU(self): self.CPU.test()
    def TestMMU(self): self.MMU.test()
    def TestMotherboard(self): self.Motherboard.test()

@singleton 
class UltraRichie(AbstractRichie):
    def __init__(self):
        super(AbstractRichie, self).__init__()
        self.CPU = CPUOfUltraRichie()
        self.MMU = MMUOfUltraRichie()
        self.Motherboard = MotherboardOfUltraRichie()
    def TestCPU(self): self.CPU.test()
    def TestMMU(self): self.MMU.test()
    def TestMotherboard(self): self.Motherboard.test()

# abstract product
class CPU(object):
    def test(self): pass
class MMU(object):
    def test(self): pass
class Motherboard(object):
    def test(self): pass

# concrete product
class CPUOfRichie(CPU):
    def __init__(self): super(CPU, self).__init__()
    def test(self): print("Test on Richie's CPU")
class CPUOfAdvancedRichie(CPU):
    def __init__(self): super(CPU, self).__init__()
    def test(self): print("Test on Advanced Richie's CPU")
class CPUOfSuperRichie(CPU):
    def __init__(self): super(CPU, self).__init__()
    def test(self): print("Test on Super Richie's CPU")
class CPUOfUltraRichie(CPU):
    def __init__(self): super(CPU, self).__init__()
    def test(self): print("Test on Ultra Richie's CPU")

class MMUOfRichie(MMU):
    def __init__(self): super(MMU, self).__init__()
    def test(self): print("Test on Richie's MMU")
class MMUOfAdvancedRichie(MMU):
    def __init__(self): super(MMU, self).__init__()
    def test(self): print("Test on Advanced Richie's MMU")
class MMUOfSuperRichie(MMU):
    def __init__(self): super(MMU, self).__init__()
    def test(self): print("Test on Super Richie's MMU")
class MMUOfUltraRichie(MMU):
    def __init__(self): super(MMU, self).__init__()
    def test(self): print("Test on Ultra Richie's MMU")

class MotherboardOfRichie(Motherboard):
    def __init__(self): super(Motherboard, self).__init__()
    def test(self): print("Test on Richie's Motherboard")
class MotherboardOfAdvancedRichie(Motherboard):
    def __init__(self): super(Motherboard, self).__init__()    
    def test(self): print("Test on Advanced Richie's Motherboard")
class MotherboardOfSuperRichie(Motherboard):
    def __init__(self): super(Motherboard, self).__init__()    
    def test(self): print("Test on Super Richie's Motherboard")
class MotherboardOfUltraRichie(Motherboard):
    def __init__(self): super(Motherboard, self).__init__()
    def test(self): print("Test on Ultra Richie's Motherboard")



if __name__ == '__main__':
    print("---Start Diagnostics on Different Architectures---")
    ConcreteRichie = AbstractRichie.GetConcreteRichie("RICHIE")
    ConcreteRichie.TestCPU()
    ConcreteRichie.TestMMU()
    ConcreteRichie.TestMotherboard()

    ConcreteRichie = AbstractRichie.GetConcreteRichie("ADVANCEDRICHIE")
    ConcreteRichie.TestCPU()
    ConcreteRichie.TestMMU()
    ConcreteRichie.TestMotherboard()

    ConcreteRichie = AbstractRichie.GetConcreteRichie("SUPERRICHIE")
    ConcreteRichie.TestCPU()
    ConcreteRichie.TestMMU()
    ConcreteRichie.TestMotherboard()

    ConcreteRichie = AbstractRichie.GetConcreteRichie("ULTRARICHIE")
    ConcreteRichie.TestCPU()
    ConcreteRichie.TestMMU()
    ConcreteRichie.TestMotherboard()

    print("\n---Test Singleton---\n"
          +"Compare two ConcreteRichies instances created consecutively with same parameter, \"True\" means these two ConcreteRichies are the same one.")
    one = AbstractRichie.GetConcreteRichie("RICHIE")
    two = AbstractRichie.GetConcreteRichie("RICHIE")
    print "Test Singleton Result: ", one == two 
